Incorporate Go, Dockerfile, and Shell lint check

tstromberg opened the pull request:
Here is the start of trying to get a standardized set of lint checks added within our Go codebases. 

While some of the lint checks are arbitrary, the consistency will make things more readable.

* Adds a base golangci-lint configuration
* Adds 'lint' makefile target that runs Go, Dockerfile, and Shell linters (the latter two only warn at the moment)
* Fix the Go lint issues raised by the base configuration
* Enables Github action to use the aforementioned `make lint` rule.
* Upgrade `pkg/errors` which implements `errors.Is` (similar to the new Go `errors` package)

mmlb commented on cmd/cacherc/cmd/all.go:
```diff
@@ -4,6 +4,7 @@ package cmd
 
 import (
 	"context"
+	"errors"
```
pkg/errors has an `Is` func, is there a reason why you didn't go with that? I can only imagine it being even more awkward to mix errors and pkg/errors uses

tstromberg replied:
I looked for one, but it didn't exist in the version of `pkg/errors` used by cacher. I've upgraded it and addressed this awkwardness. Thanks!

tobert commented on client/main.go:
```diff
@@ -15,54 +15,27 @@ import (
-
-		return fmt.Sprintf("%s:%d", strings.TrimSuffix(addrs[0].Target, "."), addrs[0].Port), nil
-	}
-	var securityOption grpc.DialOption
```
Oh fun, I did this same change in #24. Whichever one merges first the other will have to fix up. I don't mind either way.

tobert commented:
Does `make lint` execute with `make test`? If something will fail a CI run it should fail the local test too imo. This has bitten me on a few repos lately.

tstromberg commented:
> Does `make lint` execute with `make test`? If something will fail a CI run it should fail the local test too imo. This has bitten me on a few repos lately.

Not yet - but perhaps it should. Any thoughts, @mmlb ?

tstromberg commented:
The behavior has now changed so that `make test` now runs the lint checks by default. There is a new `make test-only` rule that preserves the old behavior.

I've also adjusted the CI to invoke the appropriate targets.

mmlb commented on Makefile:
```diff
@@ -16,8 +16,41 @@ protos/cacher/cacher.pb.go: protos/cacher/cacher.proto
 
 run: ${binaries}
 	docker-compose up --build server
+
+GOLINT_VERSION ?= v1.41.1
+HADOLINT_VERSION ?= v2.6.1
+SHELLCHECK_VERSION ?= v0.7.2
+OS := $(shell uname)
+LOWER_OS  = $(shell echo $OS | tr '[:upper:]' '[:lower:]')
+ARCH := $(shell uname -m)
+
+# TODO(tstromberg): Change hadolint/shell checks to be fatal once issues can be addressed
+lint: out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck out/linters/hadolint-$(HADOLINT_VERSION) out/linters/golangci-lint-$(GOLINT_VERSION)
+	out/linters/golangci-lint-$(GOLINT_VERSION) run
+	out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck $(shell find . -name "*.sh") || true
+	out/linters/hadolint-$(HADOLINT_VERSION) $(shell find . -name "*Dockerfile") || true
+
+# Shell script linter
+out/linters/shellcheck-$(SHELLCHECK_VERSION)/shellcheck:
+	mkdir -p out/linters
+	curl -sfL https://github.com/koalaman/shellcheck/releases/download/v0.7.2/shellcheck-$(SHELLCHECK_VERSION).$(OS).$(ARCH).tar.xz | tar -C out/linters -xJf -
+
+# Dockerfile linter
+out/linters/hadolint-$(HADOLINT_VERSION):
+	mkdir -p out/linters
+	curl -sfL https://github.com/hadolint/hadolint/releases/download/v2.6.1/hadolint-$(OS)-$(ARCH) > out/linters/hadolint-$(HADOLINT_VERSION)
+	chmod u+x out/linters/hadolint-$(HADOLINT_VERSION)
+
+# Go linter
+out/linters/golangci-lint-$(GOLINT_VERSION):
+	mkdir -p out/linters
+	curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b out/linters $(GOLINT_VERSION)
+	mv out/linters/golangci-lint out/linters/golangci-lint-$(GOLINT_VERSION)
```
stuff like this is what we use `nix-shell` for, have you tried that out? I'd be happy to pair up to get this to use nix-shell instead of fetching linters like this.

tstromberg replied:
> hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.

Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.

If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate

If the latter: I do not intend on splitting this PR into 46 PR's.

mmlb replied:
> > hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.
> 
> Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.

Both.

> 
> If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate
> 
> If the latter: I do not intend on splitting this PR into 46 PR's.

Definitely not 46 PRs! I was looking for 46 (probably less) commits. I'd wager most linter modules have no affect to the code base so those could be enabled right away. Then one commit per linter module so we see what it actually does. We'd be able to go through them in this one PR.

tstromberg replied:
I have not tried it, but so long as `make lint` works and installs all of its dependencies, I do not really care what tool is used underneath, assuming the tool versions are recent.

mmlb replied:
lets get this PR as is and I'll follow up with nixifying instead of fetching like this.

tstromberg replied:
OK. I'll say I am a little bit skeptical on whether nix-shell will be an improvement as you will still need to fetch and install nix-shell as part of the Makefile rule.

mmlb commented on .golangci.yml:
```diff
@@ -0,0 +1,67 @@
+    - typecheck
+    - unused
+    - varcheck
+    - dogsled
+    - gocritic
+    - gofmt
+    - goimports
+    - misspell
+    - nakedret
+    - stylecheck
+    - unconvert
+    - unparam
+    - revive
+    - ifshort
+    - whitespace
+    - wastedassign
+    - asciicheck
+    - bodyclose
+    - durationcheck
+    - errorlint
+    - exhaustive
+    - exportloopref
+    - forcetypeassert
+    - gocyclo
+    - godot
+    - goheader
+    - goprintffuncname
+    - importas
+    - makezero
+    - nestif
+    - nilerr
+    - nolintlint
+    - prealloc
+    - predeclared
+    - rowserrcheck
+    - sqlclosecheck
+    - thelper
+    - tparallel
+    - cyclop
+    - wsl
```
sort please

tstromberg replied:
done, but why?

mmlb replied:
thanks, large unsorted lists are an eyesore and feel like they're just a bit heavier. I know its a sucky request but it definitely helps me.

mmlb reviewed (commented):
hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.

mmlb commented:
> > Does `make lint` execute with `make test`? If something will fail a CI run it should fail the local test too imo. This has bitten me on a few repos lately.
> 
> Not yet - but perhaps it should. Any thoughts, @mmlb ?

I do like having the different actions be separate, tests and lint are distinct while developing the code. How about `make test`, `make lint`, `make ci` (or similar). Or maybe something else like `make (go|unit)-tests` and `make test` == what is done in ci and includes *-tests + lint?

tstromberg commented:
> > > Does `make lint` execute with `make test`? If something will fail a CI run it should fail the local test too imo. This has bitten me on a few repos lately.
> > 
> > 
> > Not yet - but perhaps it should. Any thoughts, @mmlb ?
> 
> I do like having the different actions be separate, tests and lint are distinct while developing the code. How about `make test`, `make lint`, `make ci` (or similar). Or maybe something else like `make (go|unit)-tests` and `make test` == what is done in ci and includes *-tests + lint?

That's basically what this PR does now:

* `make test`: what is done in CI
* `make tests-only`: unit/integration tests
* `make lint`: just lint checks

Is that OK?

tstromberg commented:
> hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.

Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.

If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate

If the latter: I do not intend on splitting this PR into 46 PR's.

mmlb commented:
> > > > Does `make lint` execute with `make test`? If something will fail a CI run it should fail the local test too imo. This has bitten me on a few repos lately.
> > > 
> > > 
> > > Not yet - but perhaps it should. Any thoughts, @mmlb ?
> > 
> > 
> > I do like having the different actions be separate, tests and lint are distinct while developing the code. How about `make test`, `make lint`, `make ci` (or similar). Or maybe something else like `make (go|unit)-tests` and `make test` == what is done in ci and includes *-tests + lint?
> 
> That's basically what this PR does now:
> 
>     * `make test`: what is done in CI
> 
>     * `make tests-only`: unit/integration tests
> 
>     * `make lint`: just lint checks
> 
> 
> Is that OK?

Yep sorry, I was thinking of this in the context of `make ci` doing all the things, then pivoted mid thought but kept on writing :D

mmlb commented:
(going to dupe the reply, somehow my comment was attached to a different thread...)

> > hmm this is a pretty huge PR and it has a few commits that do a lot of things. It'd be great if you could re-work so only one linter is enabled per-commit so. This would help review the effects of each linter being enabled.
> 
> Do you mean per-linter (golang, hadolint, shellcheck), or do you mean per golang linter module (the ~46 modules) we enable.
> 

Both.

> If the former: this PR only introduces changes to our code-base necessary to pass golangci-lint, but does demonstrate
> 
> If the latter: I do not intend on splitting this PR into 46 PR's.

Definitely not 46 PRs! I was looking for 46 (probably less) commits. I'd wager most linter modules have no affect to the code base so those could be enabled right away. Then one commit per linter module so we see what it actually does. We'd be able to go through them in this one PR.
...

That said, I'm a ok if you *don't* want to do that for this PR since re-working the commits is going to be pretty tedious. It *would* be good when updating the other repos though.

tstromberg commented:
I don't know if I'm ready to agree to one commit per linter, but I will promise that future PR's will be less noisy, and specifically include which lint errors are being fixed. This one got a bit out of hand as I went back and forth between whether or not I cared about the `wsl` linter results.
